src/Tools/DynamoInstallDetective/ProductLookUp.cs

```diff
@@ -9,6 +9,118 @@
 
 namespace DynamoInstallDetective
 {
+    // Utility class for interacting with the windows registry.
+    internal static class RegUtils
+    {
+        // Utility class to enable/disable registry caching within a scope.
+        class RegistryCacher : IDisposable
+        {
+            public RegistryCacher()
+            {
+                EnableCache();
+            }
+
+            public void Dispose()
+            {
+                DisableCache();
+            }
+        }
+
+        public static string REG_KEY64 = @"SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\";
+        public static string REG_KEY32 = @"SOFTWARE\Wow6432Node\Microsoft\Windows\CurrentVersion\Uninstall\";
+
+        // Cached data from windows registry. Consists of a dictionary that maps The product name/code to Display Name (Tuple.item1) and Install Path(Tuple.item2)
+        // Dictionary<ProductCode, (DisplayName, InstallPath)>
+        private static Dictionary<string, Tuple<string, string>> cachedRecords;
+        private static bool cacheEnabled = false;
+
+        private static readonly object mutex = new object();
+
+        public static RegistryKey OpenKey(string key)
+        {
+            var regKey = RegistryKey.OpenBaseKey(RegistryHive.LocalMachine, RegistryView.Registry64);
+            return regKey.OpenSubKey(key);
+        }
+        public static string GetInstallLocation(RegistryKey key)
+        {
+            if (key != null)
+                return key.GetValue("InstallLocation") as string;
+
+            return string.Empty;
+        }
+
+        public static string GetDisplayName(RegistryKey key)
+        {
+            if (key != null)
+                return key.GetValue("DisplayName") as string;
+
+            return string.Empty;
+        }
+
+        // Returns all the products registered under "SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\" that have a valid DisplayName and an InstallLocation.
+        public static Dictionary<string, Tuple<string, string>> GetInstalledProducts()
```

Comment:
this is the main change in this PR. 
We now iterate through all the registry subkeys and we return only those that have a DIsplayName and an InstalledLocation.

We will find asm location in by matching the DisplayName in the above records